src/libslic3r/PerimeterGenerator.cpp

```diff
@@ -1923,29 +1932,44 @@ void PerimeterGenerator::process_arachne()
         std::vector<Arachne::VariableWidthLines> out_shell;
         ExPolygons top_fills;
         ExPolygons fill_clip;
-        if (loop_number > 0 && config->only_one_wall_top && !surface.is_bridge() && this->upper_slices != nullptr) {
-            // Check if current layer has surfaces that are not covered by upper layer (i.e., top surfaces)
+
+        // Check if we're on a top surface, and make adjustments where needed
+        if (!surface.is_bridge() && !is_topmost_layer) {
             ExPolygons non_top_polygons;
-            this->split_top_surfaces(last, top_fills, non_top_polygons, fill_clip);
+            // Temporary storage, in the event all we need to do is set is_top_or_bottom_layer
+            ExPolygons top_fills_tmp;
+            ExPolygons fill_clip_tmp;
+            // Check if current layer has surfaces that are not covered by upper layer (i.e., top surfaces)
+            this->split_top_surfaces(last, top_fills_tmp, non_top_polygons, fill_clip_tmp);
 
-            if (top_fills.empty()) {
+            if (top_fills_tmp.empty()) {
                 // No top surfaces, no special handling needed
             } else {
-                // First we slice the outer shell
-                Polygons last_p = to_polygons(last);
-                Arachne::WallToolPaths wallToolPaths(last_p, bead_width_0, perimeter_spacing, coord_t(1),
-                                                     wall_0_inset, layer_height, input_params);
-                out_shell = wallToolPaths.getToolPaths();
-                // Make sure infill not overlap with wall
-                top_fills = intersection_ex(top_fills, wallToolPaths.getInnerContour());
-
-                if (!top_fills.empty()) {
-                    // Then get the inner part that needs more walls
-                    last = intersection_ex(non_top_polygons, wallToolPaths.getInnerContour());
-                    loop_number--;
-                } else {
-                    // Give up the outer shell because we don't have any meaningful top surface
-                    out_shell.clear();
+                // Adjust arachne input params to prevent removal of larger short walls, which could lead to gaps
+                input_params.is_top_or_bottom_layer = true;
```

Comment:
Shall we make a local copy of the input_params and only change value for the copy?
Since we have split the surface into multiple regions at this point, non top surface shouldn't been affected at Ln1979.